From f12a2cee9efadce4254d15a6b7fe59f1e1cc0f10 Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Thu, 28 Nov 2024 09:03:12 +0100
Subject: [PATCH 1395/1695] clk: stm32mp21: configure STGEN flexgen in .enable
 ops

STGEN flexgen is skipped during RCC probe to prevent misalignment
between stgen_clk frequency and STGEN register.
In .enable ops, reparent flexgen33 clock to the clock source found in
the device tree.
This allow the driver stm32_stgen driver to align the STGEN
configuration just after the configuration of the STGEN flexgen.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8e7a49f96832ef7cf09bd761369775869344077f clk: stm32mp21: configure STGEN flexgen in .enable ops]
Change-Id: I902107056c7174ff866d6f9c7b897b4d4cb7a121
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/416439
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/426777
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/clk/clk-stm32mp21.c | 73 ++++++++++++++++++++++++++++++--
 1 file changed, 69 insertions(+), 4 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index 1c007f953..51d32e4fa 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -1950,6 +1950,41 @@ static int wait_xbar_sts(uint16_t channel)
 	return 0;
 }
 
+static TEE_Result flexclkgen_search_config(uint16_t channel,
+					   unsigned int *clk_src,
+					   unsigned int *prediv,
+					   unsigned int *findiv)
+{
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	struct stm32_clk_platdata *pdata = priv->pdata;
+	unsigned int flex_id = U(0);
+	uint32_t dt_cfg = U(0);
+	uint32_t i = U(0);
+
+	assert(clk_src && prediv && findiv);
+
+	/*
+	 * pdata->flexgen is the array of all the flexgen configuration from
+	 * the device tree.
+	 * The binding does not enforce the description of all flexgen nor
+	 * the order it which they are listed.
+	 */
+	for (i = 0; i < pdata->nflexgen; i++) {
+		dt_cfg = pdata->flexgen[i];
+
+		flex_id = (dt_cfg & FLEX_ID_MASK) >> FLEX_ID_SHIFT;
+		if (flex_id == channel) {
+			*clk_src = (dt_cfg & FLEX_SEL_MASK) >> FLEX_SEL_SHIFT;
+			*prediv = (dt_cfg & FLEX_PDIV_MASK) >> FLEX_PDIV_SHIFT;
+			*findiv = (dt_cfg & FLEX_FDIV_MASK) >> FLEX_FDIV_SHIFT;
+
+			return TEE_SUCCESS;
+		}
+	}
+
+	return TEE_ERROR_ITEM_NOT_FOUND;
+}
+
 static void flexclkgen_config_channel(uint16_t channel, unsigned int clk_src,
 				      unsigned int prediv, unsigned int findiv)
 {
@@ -2011,7 +2046,7 @@ static int stm32mp2_clk_flexgen_configure(struct clk_stm32_priv *priv)
 		channel = (cmd_data & FLEX_ID_MASK) >> FLEX_ID_SHIFT;
 
 		/*
-		 * Skip ck_ker_stgen configuration, will be done by
+		 * Skip ck_ker_stgen configuration, will be done when enable by
 		 * stgen driver.
 		 */
 		if (channel == FLEX_STGEN)
@@ -2862,11 +2897,41 @@ static TEE_Result clk_stm32_flexgen_enable(struct clk *clk)
 {
 	struct clk_stm32_flexgen_cfg *cfg = clk->priv;
 	uintptr_t rcc_base = clk_stm32_get_rcc_base();
+	TEE_Result ret = TEE_ERROR_GENERIC;
 	uint8_t channel = cfg->flex_id;
 
-	if (stm32_rcc_has_access_by_id(cfg->flex_id))
-		io_setbits32(rcc_base + RCC_FINDIV0CFGR + (0x4 * channel),
-			     RCC_FINDIV0CFGR_FINDIV0EN);
+	if (!stm32_rcc_has_access_by_id(channel))
+		return TEE_SUCCESS;
+
+	/*
+	 * Configure flexgen of STGEN since it has been skipped during
+	 * flexgen configuration.
+	 */
+	if (channel == FLEX_STGEN) {
+		struct clk *stgen_src = NULL;
+		unsigned int clk_src = U(0);
+		unsigned int pdiv = U(0);
+		unsigned int fdiv = U(0);
+
+		ret = flexclkgen_search_config(channel, &clk_src, &pdiv, &fdiv);
+		if (ret) {
+			EMSG("Error %#x when getting STGEN flexgen conf", ret);
+			return ret;
+		}
+
+		flexclkgen_config_channel(channel, clk_src, pdiv, fdiv);
+
+		/* Update parent */
+		stgen_src = clk_get_parent_by_index(clk, clk_src);
+		ret = clk_reparent(clk, stgen_src);
+		if (ret) {
+			EMSG("Failed to configured the STGEN flexgen");
+			return ret;
+		}
+	}
+
+	io_setbits32(rcc_base + RCC_FINDIV0CFGR + (0x4 * channel),
+		     RCC_FINDIV0CFGR_FINDIV0EN);
 
 	return TEE_SUCCESS;
 }
-- 
2.47.2

